File: C++/tests/TestEntity.cpp
#include <assert.h>
#include <Entity.h>

using namespace std;

EntityCore Core;

// Component Creation
Component<float> LEFT 	= Core.newComponent("left", 0.0f);
Component<float> RIGHT 	= Core.newComponent("right", 0.0f);
Component<float> SPEED 	= Core.newComponent("speed", 0.0f);

// Controller, View, Dynamic Component, & EntityType Creation
ComponentGet<float> CENTER 	= Core.newComponentGet<float>("center", {LEFT.id, RIGHT.id},
	[](Entity &e) -> float {
		return ( e(LEFT) + e(RIGHT) ) * 0.5f;
	}
);

size_t MOTION = Core.addController({LEFT.id, RIGHT.id, SPEED.id},
	[](Entity &e, void *updateState) {
		float dt = *((float*)updateState);
		float s = e(SPEED);

		e(LEFT) += s * dt;
		e(RIGHT) += s * dt;
	}
);

size_t EXTENT_VIEW = Core.addView({LEFT.id, RIGHT.id}, 
	[](Entity &e, void *drawState) {
		string graphics = *((string*)drawState);
		cout << "Drawing extent at {" << e(LEFT) << "->" << e(RIGHT) << "} with " << graphics << "." << endl;
	}
);

EntityType* EXTENT = Core.newEntityType({LEFT.id, RIGHT.id}, {MOTION}, EXTENT_VIEW);

void testConstructorId()
{
   	cout << "Running " << __func__ << "() ... ";

   	Entity e(EXTENT);

   	assert( e.has(LEFT) );
   	assert( e.has(RIGHT) );
   	assert( e.hasController(MOTION) );

	cout << "Pass" << endl;
}

void testConstructorReference()
{
  	cout << "Running " << __func__ << "() ... ";	

	EntityType *type = Core.getEntityType(EXTENT->getId());

	Entity e(type);

   	assert( e.has(LEFT) );
   	assert( e.has(RIGHT) );
   	assert( e.hasController(MOTION) );
   	assert( e.getEntityType() == type );

	cout << "Pass" << endl;
}

void testCustomEntity()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core);

	assert( e.isCustom() );

	assert(!e.has(LEFT) );

	e.add(LEFT);

	assert( e.has(LEFT) );

	e.set(LEFT, 3.0f);

	assert( e.get(LEFT) == 3.0f );

	cout << "Pass" << endl;
}

void testCustomEntityDefined()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id}, {}, View::NONE);

	assert( e.has(LEFT) );
	assert(!e.has(RIGHT) );
	assert(!e.hasController(MOTION) );

	cout << "Pass" << endl;
}

void testSetMethod()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	e(LEFT, 3.5f);

	assert( e.get(LEFT) == 3.5f );

	e(LEFT) = 3.467f;

	assert( e.get(LEFT) == 3.467f );

	cout << "Pass" << endl;
}

void testToString()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	e(LEFT) = 3.0f;
	e(RIGHT) = 5.5f;

	cout << e << " ";

	cout << "Pass" << endl;
}

void testPtr()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	float* l = e.ptr(LEFT);
	float* r = e.ptr(RIGHT);

	assert( l != nullptr );
	assert( r != nullptr );

	assert( *l != 3.5f );
	assert( *r !=-2.0f );

	*l = 3.5f;
	*r =-2.0f;

	assert( *l == 3.5f );
	assert( *r ==-2.0f );

	assert( e.get(LEFT) == 3.5f );
	assert( e.get(RIGHT) ==-2.0f );

	cout << "Pass" << endl;
}

void testPtrSafe()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id, RIGHT.id}, {}, View::NONE);

	assert( e.has(LEFT) );
	assert( e.has(RIGHT) );
	assert(!e.has(SPEED) );

	assert( e.ptrs(LEFT) != nullptr );
	assert( e.ptrs(RIGHT) != nullptr );
	assert( e.ptrs(SPEED) == nullptr );

	cout << "Pass" << endl;
}

void testGet()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	e.get(LEFT) = 3.0f;
	e.get(RIGHT) = 5.0f;

	assert( e.get(LEFT) == 3.0f );
	assert( e.get(RIGHT) == 5.0f );

	cout << "Pass" << endl;
}

void testGetDynamic()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	e(LEFT) = 2.0f;
	e(RIGHT) = 3.0f;

	float center = e(CENTER);

	assert( center == 2.5f );

	cout << "Pass" << endl;
}

void testGetDynamicMissing()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id}, {}, View::NONE);

	assert( e(LEFT, 2.0f) );
	assert(!e(RIGHT, 3.0f) );

	float center = e.gets(CENTER);

	assert( center == 0.0f );

	cout << "Pass" << endl;
}

void testGetSafe()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id}, {}, View::NONE);

	e(LEFT, 2.0f);
	e(RIGHT, 3.5f);

	assert( e.gets(LEFT, 0.0f) == 2.0f );
	assert( e.gets(RIGHT, 0.0f) == 0.0f );

	cout << "Pass" << endl;
}

void testGrab()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id}, {}, View::NONE);

	e(LEFT, 2.0f);
	e(RIGHT, 3.5f);

	float left = 0.0f;
	assert( e.grab(LEFT, &left) );
	assert( left == 2.0f );

	float right = 0.0f;
	assert(!e.grab(RIGHT, &right) );
	assert( right == 0.0f );

	cout << "Pass" << endl;
}

void testHasComponents()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id, RIGHT.id}, {}, View::NONE);

	assert( e.has({LEFT.id}) );
	assert( e.has({RIGHT.id}) );
	assert( e.has({RIGHT.id, LEFT.id}) );
	assert(!e.has({SPEED.id}) );
	assert(!e.has({LEFT.id,SPEED.id}) );

	cout << "Pass" << endl;
}

void testHasController()
{
	cout << "Running " << __func__ << "() ... ";

	Entity a(&Core, {LEFT.id}, {MOTION}, View::NONE);

	assert( a.hasController(MOTION) );

	Entity b(&Core, {LEFT.id}, {}, View::NONE);

	assert(!b.hasController(MOTION) );

	cout << "Pass" << endl;
}

void testAdd()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {LEFT.id}, {MOTION}, View::NONE);

	assert( e.has(LEFT) );
	assert(!e.has(RIGHT) );

	e(LEFT, 2.0f);

	assert( e.add(RIGHT, 3.4f) );

	assert( e.has(RIGHT) );
	assert( e.gets(LEFT, 0.0f) == 2.0f );
	assert( e.gets(RIGHT, 0.0f) == 3.4f );
	assert( e.isCustom() );

	assert(!e.add(LEFT) );
	assert(!e.add(RIGHT) );
	assert( e.add(SPEED) );

	e(SPEED, 3.0f);

	assert( e.gets(SPEED, 0.0f) == 3.0f );

	cout << "Pass" << endl;
}

void testExpire()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	assert(!e.isExpired() );

	e.expire();

	assert( e.isExpired() );

	cout << "Pass" << endl;
}

void testVisible()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	assert( e.isVisible() );

	e.hide();

	assert(!e.isVisible() );

	e.show();

	assert( e.isVisible() );

	cout << "Pass" << endl;
}

// testVisibleDraw() 
Component<int> DRAWS = Core.newComponent("draws", 0);

size_t DRAWS_VIEW = Core.addView({DRAWS.id}, 
	[](Entity &e, void *drawState) {
		e(DRAWS)++;
	}
);

EntityType* DRAWS_TYPE = Core.newEntityType({DRAWS.id}, {}, DRAWS_VIEW);

void testVisibleDraw()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(DRAWS_TYPE);

	assert( e.get(DRAWS) == 0 );

	e.draw(nullptr);

	assert( e.get(DRAWS) == 1 );

	e.hide();
	e.draw(nullptr);

	assert( e.get(DRAWS) == 1 );

	e.show();
	e.draw(nullptr);

	assert( e.get(DRAWS) == 2 );

	cout << "Pass" << endl;
}

void testSetView()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(DRAWS_TYPE);

	assert( e.hasView() );
	assert( e.getView() == DRAWS_VIEW );
	assert(!e.isCustom() );

	e.setView( View::NONE );

	assert( e.isCustom() );
	assert(!e.hasView() );
	assert( e.getView() == View::NONE );
	
	cout << "Pass" << endl;
}

void testEnabled()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);

	assert( e.isEnabled() );

	e.disable();

	assert(!e.isEnabled() );

	e.enable();

	assert( e.isEnabled() );

	cout << "Pass" << endl;
}

// testVisibleDraw() 
Component<int> UPDATES = Core.newComponent("draws", 0);

size_t UPDATES_CONTROLLER = Core.addController({UPDATES.id}, 
	[] (Entity& e, void *updateState) {
		e(UPDATES)++;
	}
);

EntityType* UPDATES_TYPE = Core.newEntityType({UPDATES.id}, {UPDATES_CONTROLLER}, View::NONE);

void testEnabledUpdate()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(UPDATES_TYPE);

	assert( e.get(UPDATES) == 0 );

	e.update(nullptr);

	assert( e.get(UPDATES) == 1 );

	e.disable();
	e.update(nullptr);

	assert( e.get(UPDATES) == 1 );

	e.enable();
	e.update(nullptr);

	assert( e.get(UPDATES) == 2 );

	cout << "Pass" << endl;
}

void testControllerEnabled()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(UPDATES_TYPE);

	assert( e.get(UPDATES) == 0 );

	e.update(nullptr);

	assert( e.get(UPDATES) == 1 );

	assert( e.isControllerEnabled(UPDATES_CONTROLLER) );

	e.disable( UPDATES_CONTROLLER );

	assert( !e.isControllerEnabled(UPDATES_CONTROLLER) );

	e.update(nullptr);

	assert( e.get(UPDATES) == 1 );

	e.enable( UPDATES_CONTROLLER );

	assert( e.isControllerEnabled(UPDATES_CONTROLLER) );

	e.update(nullptr);

	assert( e.get(UPDATES) == 2 );

	cout << "Pass" << endl;
}

void testAddController()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(&Core, {UPDATES.id}, {}, View::NONE);

	assert(!e.hasController(UPDATES_CONTROLLER) );
	assert( e.get(UPDATES) == 0 );

	e.update(nullptr);

	assert( e.get(UPDATES) == 0 );

	e.addController(UPDATES_CONTROLLER);
	assert( e.hasController(UPDATES_CONTROLLER) );

	e.update(nullptr);

	assert( e.get(UPDATES) == 1 );

	cout << "Pass" << endl;
}

void testClone()
{
	cout << "Running " << __func__ << "() ... ";

	Entity *a = new Entity(EXTENT);
	a->set(LEFT, 2.0f);
	a->set(RIGHT, 3.7f);

	Entity *b = a->clone();

	assert( a != b );
	assert( b->get(LEFT) == 2.0f );
	assert( b->get(RIGHT) == 3.7f );

	a->set(LEFT, -3.5f);
	a->set(RIGHT, 5.6f);

	assert( b->get(LEFT) == 2.0f );
	assert( b->get(RIGHT) == 3.7f );

	delete b;
	delete a;

	cout << "Pass" << endl;
}

void testSetMap()
{
	cout << "Running " << __func__ << "() ... ";

	Entity e(EXTENT);
	e.set({{LEFT.id,3.0f},{RIGHT.id,4.5f}});

	assert( e.get(LEFT) == 3.0f );
	assert( e.get(RIGHT) == 4.5f );
	
	cout << "Pass" << endl;
}

void testEquals()
{
	cout << "Running " << __func__ << "() ... ";

	Entity a(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.5f}});
	Entity b(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.5f}});
	Entity c(EXTENT, {{LEFT.id,3.1f},{RIGHT.id,4.5f}});
	Entity d(&Core);
	Entity e(&Core);

	assert( a == b );
	assert( a != c );
	assert( a != d );
	assert( a != e );

	assert( b != c );
	assert( b != d );
	assert( b != e );

	assert( c != d );
	assert( c != e );

	assert( d != e ); // Their types are different, maybe I should do more than testing address?

	cout << "Pass" << endl;
}

void testHashCode()
{
	cout << "Running " << __func__ << "() ... ";

	Entity a(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.5f}});
	Entity b(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.5f}});
	Entity c(EXTENT, {{LEFT.id,3.1f},{RIGHT.id,4.5f}});
	Entity d(&Core);
	Entity e(&Core);

	assert( a.hashCode() == b.hashCode() );
	assert( a.hashCode() != c.hashCode() );
	assert( a.hashCode() != d.hashCode() );
	assert( a.hashCode() != e.hashCode() );

	assert( b.hashCode() != c.hashCode() );
	assert( b.hashCode() != d.hashCode() );
	assert( b.hashCode() != e.hashCode() );

	assert( c.hashCode() != d.hashCode() );
	assert( c.hashCode() != e.hashCode() );

	assert( d.hashCode() == e.hashCode() );
	
	cout << "Pass" << endl;
}

void testCompareTo()
{
	cout << "Running " << __func__ << "() ... ";

	Entity a(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.5f}});
	Entity b(EXTENT, {{LEFT.id,3.1f},{RIGHT.id,4.5f}});
	Entity c(EXTENT, {{LEFT.id,3.0f},{RIGHT.id,4.6f}});
	Entity d(&Core);
	
	assert( a < b );
	assert( a < b );
	assert( a > d ); // Entities without type go first
	assert( b > c );
	assert( b > d );
	assert( c > d );

	cout << "Pass" << endl 	;
}

int main()
{
	cout << "Starting " << __FILE__ << "..." << endl;

	testConstructorId();
	testConstructorReference();
	testCustomEntity();
	testCustomEntityDefined();
	testSetMethod();
	testToString();
	testPtr();
	testPtrSafe();
	testGet();
	testGetDynamic();
	testGetDynamicMissing();
	testGetSafe();
	testGrab();
	testHasComponents();
	testHasController();
	testAdd();
	testExpire();
	testVisible();
	testVisibleDraw();
	testSetView();
	testEnabled();
	testEnabledUpdate();
	testControllerEnabled();
	testAddController();
	testClone();
	testSetMap();
	testEquals();
	testHashCode();
	testCompareTo();

   	cout << __FILE__ << " has ran successfully." << endl;

	return 0;
}

File: Java/test/org/magnos/entity/test/TestEntity.java

package org.magnos.entity.test;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertNotSame;
import static org.junit.Assert.assertNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertTrue;

import java.util.concurrent.atomic.AtomicBoolean;

import org.junit.AfterClass;
import org.junit.Test;
import org.magnos.entity.Component;
import org.magnos.entity.Components;
import org.magnos.entity.Control;
import org.magnos.entity.Controller;
import org.magnos.entity.Controllers;
import org.magnos.entity.DynamicValue;
import org.magnos.entity.Entity;
import org.magnos.entity.EntityCore;
import org.magnos.entity.Renderer;
import org.magnos.entity.Template;
import org.magnos.entity.View;
import org.magnos.entity.vals.FloatVal;
import org.magnos.entity.vals.IntVal;


public class TestEntity
{

   public static final float EPSILON = 0.000001f;

   @AfterClass
   public static void afterTest()
   {
      EntityCore.clear();
   }

   // @formatter:off
   static DynamicValue<FloatVal> DYNAMIC_CENTER = new DynamicValue<FloatVal>() {
      public FloatVal get( Entity e ) {
         return take( e, new FloatVal() );
      }
      public void set( Entity e, FloatVal value ) {
         FloatVal r = e.get( RIGHT );
         FloatVal l = e.get( LEFT );
         float halfwidth = (r.v - l.v) * 0.5f;
         r.v = value.v + halfwidth;
         l.v = value.v - halfwidth;
      }
      public FloatVal take( Entity e, FloatVal target ) {
         FloatVal r = e.get( RIGHT );
         FloatVal l = e.get( LEFT );
         target.v = (l.v + r.v) * 0.5f;
         return target;
      }
   };

   static Control CONTROL_MOTION = new Control() {
      public void control( Entity e, Object updateState ) {
         float dt = (Float)updateState;
         float dist = e.get( SPEED ).v * dt;
         e.get( LEFT ).v += dist;
         e.get( RIGHT ).v += dist;
      }
   };
   
   static Control CONTROL_UPDATES = new Control() {
      public void control( Entity e, Object updateState ) {
         e.get( UPDATE_COUNT ).v++;
      }
   };
   
   static Renderer RENDERER_EXTENT = new Renderer() {
      public void draw( Entity e, Object drawState ) {
         String graphics = drawState.toString();
         System.out.println( "Drawing extent at {" + e.get( LEFT ) + "->" + e.get( RIGHT ) + "} with " + graphics );
      }
      public void destroy( Entity e ) { }
      public Renderer create( Entity e ) { return this; }
   };
   
   static Renderer RENDERER_DRAWS = new Renderer() {
      public void draw( Entity e, Object drawState ) {
         e.get( DRAW_COUNT ).v++;
      }
      public void destroy( Entity e ) { }
      public Renderer create( Entity e ) { return this; }
   };
   
   static Component<FloatVal>   LEFT            = EntityCore.newComponent( "left", new FloatVal() );
   static Component<FloatVal>   RIGHT           = EntityCore.newComponent( "right", new FloatVal() );
   static Component<FloatVal>   SPEED           = EntityCore.newComponent( "speed", new FloatVal() );
   static Component<FloatVal>   CENTER          = EntityCore.newComponentDynamic( "center", DYNAMIC_CENTER );
   static Controller            MOTION          = EntityCore.newController( "motion", CONTROL_MOTION );
   static Controller            NETWORKING      = EntityCore.newController( "networking" );
   static View                  EXTENT_VIEW     = EntityCore.newView( "extent-view", RENDERER_EXTENT );
   static Template              EXTENT          = EntityCore.newTemplate( "extent", new Components(LEFT, RIGHT, CENTER), new Controllers(MOTION), EXTENT_VIEW );
   
   static Component<IntVal>     DRAW_COUNT      = EntityCore.newComponent( "draw-count", new IntVal() );
   static View                  DRAWS_VIEW      = EntityCore.newView( "draws-view", RENDERER_DRAWS );
   static Template              DRAWS           = EntityCore.newTemplate( "draws", new Components(DRAW_COUNT), Controllers.NONE, DRAWS_VIEW );
   
   static Component<IntVal>     UPDATE_COUNT    = EntityCore.newComponent( "update-count", new IntVal() );
   static Controller            UPDATES_CONTROL = EntityCore.newController( "updates-control", CONTROL_UPDATES );
   static Template              UPDATES         = EntityCore.newTemplate( "updates", new Components(UPDATE_COUNT), new Controllers(UPDATES_CONTROL) );
   
   // @formatter:on

   @Test
   public void testConstructor()
   {
      Entity e = new Entity( EXTENT );

      assertTrue( e.isVisible() );
      assertTrue( e.isEnabled() );
      assertFalse( e.isExpired() );
      assertFalse( e.isCustom() );

      assertTrue( e.has( LEFT ) );
      assertTrue( e.has( RIGHT ) );
      assertTrue( e.has( CENTER ) );
      assertFalse( e.has( SPEED ) );
      assertTrue( e.has( MOTION ) );
      assertTrue( e.has( EXTENT_VIEW ) );

      e.delete();
   }

   @Test
   public void testCustomEntity()
   {
      Entity e = new Entity();

      assertTrue( e.isCustom() );
      assertFalse( e.has( LEFT ) );

      e.add( LEFT );

      assertTrue( e.has( LEFT ) );

      e.get( LEFT ).v = 3.0f;

      assertEquals( 3.0f, e.get( LEFT ).v, EPSILON );

      e.delete();
   }

   @Test
   public void testCustomEntityDefined()
   {
      Entity e = new Entity( LEFT );

      assertTrue( e.has( LEFT ) );
      assertFalse( e.has( RIGHT ) );
      assertFalse( e.has( MOTION ) );

      e.delete();
   }

   @Test
   public void testToString()
   {
      // Non-custom
      Entity e0 = new Entity( EXTENT );
      e0.get( LEFT ).v = 3.0f;
      e0.get( RIGHT ).v = 5.5f;
      assertEquals( "[left=3.0,right=5.5,center=4.25]", e0.toString() );

      // Custom empty
      Entity e1 = new Entity();
      assertEquals( "[]", e1.toString() );

      // Custom non-empty
      Entity e2 = new Entity( LEFT );
      e2.get( LEFT ).v = 2.0f;
      assertEquals( "[left=2.0]", e2.toString() );

      // Custom from template
      Entity e3 = new Entity( EXTENT );
      e3.get( LEFT ).v = 3.0f;
      e3.get( RIGHT ).v = 5.5f;
      e3.grab( SPEED ).v = -1.0f;
      assertEquals( "[left=3.0,right=5.5,center=4.25,speed=-1.0]", e3.toString() );

      e0.delete();
      e1.delete();
      e2.delete();
      e3.delete();
   }

   @Test
   public void testGet()
   {
      Entity e = new Entity( EXTENT );

      FloatVal left = e.get( LEFT );
      FloatVal right = e.get( RIGHT );

      left.v = 3.0f;
      right.v = 5.5f;

      assertSame( left, e.get( LEFT ) );
      assertSame( right, e.get( RIGHT ) );
      assertEquals( 3.0f, e.get( LEFT ).v, EPSILON );
      assertEquals( 5.5f, e.get( RIGHT ).v, EPSILON );

      e.delete();
   }

   @Test
   public void testDynamic()
   {
      Entity e = new Entity( EXTENT );

      e.get( LEFT ).v = 2.0f;
      e.get( RIGHT ).v = 3.0f;

      FloatVal center = e.get( CENTER );

      assertEquals( 2.5f, center.v, EPSILON );

      e.delete();
   }

   @Test
   public void testGetSafe()
   {
      Entity e = new Entity( EXTENT );

      assertFalse( e.has( SPEED ) );

      FloatVal speed0 = e.gets( SPEED );

      assertNull( speed0 );

      FloatVal speed1 = e.gets( SPEED, new FloatVal( Float.NaN ) );

      assertNotNull( speed1 );
      assertTrue( Float.isNaN( speed1.v ) );

      e.delete();
   }

   @Test
   public void testTake()
   {
      final FloatVal left = new FloatVal( 0.0f );
      final FloatVal right = new FloatVal( 0.0f );

      Entity e = new Entity( LEFT );

      e.get( LEFT ).v = 2.0f;

      FloatVal leftTaken = e.takes( LEFT, left );
      assertSame( left, leftTaken );
      assertNotSame( left, e.gets( LEFT ) );
      assertEquals( 2.0f, left.v, EPSILON );

      FloatVal rightTaken = e.takes( RIGHT, right );
      assertNull( rightTaken );
      assertNotSame( right, e.gets( RIGHT ) );
      assertEquals( 0.0f, right.v, EPSILON );

      e.delete();
   }

   @Test
   public void testHasComponents()
   {
      Entity e = new Entity( EXTENT );

      assertTrue( e.has( LEFT ) );
      assertTrue( e.has( RIGHT ) );
      assertTrue( e.has( CENTER ) );
      assertFalse( e.has( SPEED ) );
      assertTrue( e.has( LEFT, RIGHT, CENTER ) );
      assertFalse( e.has( LEFT, RIGHT, CENTER, SPEED ) );

      e.delete();
   }

   @Test
   public void testHasController()
   {
      Entity e = new Entity( EXTENT );

      assertTrue( e.has( MOTION ) );
      assertFalse( e.has( NETWORKING ) );
      assertFalse( e.has( MOTION, NETWORKING ) );

      e.delete();
   }

   @Test
   public void testAdd()
   {
      Entity e = new Entity( new Components( LEFT ), new Controllers( MOTION ) );

      assertTrue( e.has( LEFT ) );
      assertFalse( e.has( RIGHT ) );

      e.get( LEFT ).v = 2.0f;

      assertTrue( e.add( RIGHT ) );
      assertTrue( e.has( RIGHT ) );

      e.get( RIGHT ).v = 3.4f;

      assertEquals( 2.0f, e.get( LEFT ).v, EPSILON );
      assertEquals( 3.4f, e.get( RIGHT ).v, EPSILON );
      assertTrue( e.isCustom() );

      assertFalse( e.add( LEFT ) );
      assertFalse( e.add( RIGHT ) );
      assertTrue( e.add( SPEED ) );

      e.get( SPEED ).v = 3.0f;

      assertEquals( 3.0f, e.get( SPEED ).v, EPSILON );

      e.delete();
   }

   @Test
   public void testExpireAndDelete()
   {
      assertEquals( 0, EXTENT.getInstances() );

      Entity e = new Entity( EXTENT );

      assertEquals( 1, EXTENT.getInstances() );

      assertFalse( e.isExpired() );
      assertNotNull( e.getTemplate() );
      assertNotNull( e.getRenderer() );

      e.expire();

      assertTrue( e.isExpired() );
      
      assertNotNull( e.getTemplate() );
      assertNotNull( e.getRenderer() );
      
      e.delete();
      
      assertNull( e.getTemplate() );
      assertNull( e.getRenderer() );

      assertEquals( 0, EXTENT.getInstances() );

      e.delete();

      assertTrue( e.isExpired() );
      assertNull( e.getTemplate() );
      assertNull( e.getRenderer() );

      assertEquals( 0, EXTENT.getInstances() );
   }

   @Test
   public void testVisible()
   {
      Entity e = new Entity( EXTENT );

      assertTrue( e.isVisible() );

      e.hide();

      assertFalse( e.isVisible() );

      e.show();

      assertTrue( e.isVisible() );

      e.delete();
   }

   @Test
   public void testVisibleDraw()
   {
      Entity e = new Entity( DRAWS );

      assertEquals( 0, e.get( DRAW_COUNT ).v );

      e.draw( null );

      assertEquals( 1, e.get( DRAW_COUNT ).v );

      e.hide();
      e.draw( null );

      assertEquals( 1, e.get( DRAW_COUNT ).v );

      e.show();
      e.draw( null );

      assertEquals( 2, e.get( DRAW_COUNT ).v );

      e.delete();
   }

   @Test
   public void testSetView()
   {
      Entity e = new Entity( DRAWS );

      assertTrue( e.hasView() );
      assertTrue( e.hasRenderer() );
      assertSame( DRAWS_VIEW, e.getView() );
      assertFalse( e.isCustom() );

      e.setView( null );

      assertTrue( e.isCustom() );
      assertFalse( e.hasView() );
      assertFalse( e.hasRenderer() );
      assertNull( e.getView() );
      assertNull( e.getRenderer() );

      e.delete();
   }

   @Test
   public void testSetRenderer()
   {
      Entity e = new Entity();

      assertTrue( e.isCustom() );
      assertFalse( e.hasRenderer() );
      assertNull( e.getRenderer() );

      final AtomicBoolean flagCreated = new AtomicBoolean( false );
      final AtomicBoolean flagDestroyed = new AtomicBoolean( false );

      e.setRenderer( new Renderer() {

         public Renderer create( Entity e )
         {
            flagCreated.set( true );
            return this;
         }

         public void draw( Entity e, Object drawState )
         {

         }

         public void destroy( Entity e )
         {
            flagDestroyed.set( true );
         }
      } );

      assertTrue( flagCreated.get() );
      assertFalse( flagDestroyed.get() );

      assertTrue( e.hasRenderer() );

      e.setRenderer( null );

      assertFalse( e.hasRenderer() );
      assertNull( e.getRenderer() );
      assertTrue( flagDestroyed.get() );

      e.delete();
   }

   @Test
   public void testEnabled()
   {
      Entity e = new Entity( EXTENT );

      assertTrue( e.isEnabled() );

      e.disable();

      assertFalse( e.isEnabled() );

      e.enable();

      assertTrue( e.isEnabled() );

      e.delete();
   }

   @Test
   public void testEnabledUpdate()
   {
      Entity e = new Entity( UPDATES );

      assertEquals( 0, e.get( UPDATE_COUNT ).v );

      e.update( null );

      assertEquals( 1, e.get( UPDATE_COUNT ).v );

      e.disable();
      e.update( null );

      assertEquals( 1, e.get( UPDATE_COUNT ).v );

      e.enable();
      e.update( null );

      assertEquals( 2, e.get( UPDATE_COUNT ).v );

      e.delete();
   }

   @Test
   public void testControllerEnabled()
   {
      Entity e = new Entity( UPDATES );

      assertEquals( 0, e.get( UPDATE_COUNT ).v );

      assertTrue( e.isControllerEnabled( UPDATES_CONTROL ) );

      e.update( null );

      assertEquals( 1, e.get( UPDATE_COUNT ).v );

      e.disable( UPDATES_CONTROL );
      assertFalse( e.isControllerEnabled( UPDATES_CONTROL ) );

      e.update( null );

      assertEquals( 1, e.get( UPDATE_COUNT ).v );

      e.enable( UPDATES_CONTROL );
      e.update( null );

      assertTrue( e.isControllerEnabled( UPDATES_CONTROL ) );

      assertEquals( 2, e.get( UPDATE_COUNT ).v );

      e.delete();
   }

   @Test
   public void testAddController()
   {
      Entity e = new Entity( UPDATE_COUNT );

      assertFalse( e.has( UPDATES_CONTROL ) );
      assertEquals( 0, e.get( UPDATE_COUNT ).v );
      
      e.update( null );
      
      assertEquals( 0, e.get( UPDATE_COUNT ).v );
      
      e.add( UPDATES_CONTROL );

      assertTrue( e.has( UPDATES_CONTROL ) );
      assertTrue( e.isControllerEnabled( UPDATES_CONTROL ) );
      
      e.update( null );
      
      assertEquals( 1, e.get( UPDATE_COUNT ).v );
      
      e.delete();
   }
   
   @Test
   public void testCloneDeep()
   {
      Entity e0 = new Entity( EXTENT );
      e0.get( LEFT ).v = 2.0f;
      e0.get( RIGHT ).v = 3.7f;
      e0.hide();
      
      Entity e1 = e0.clone( true );
      
      assertSame( e0.getTemplate(), e1.getTemplate() );
      assertNotSame( e0.get( LEFT ), e1.get( LEFT ) );
      assertNotSame( e0.get( RIGHT ), e1.get( RIGHT ) );
      assertFalse( e1.isVisible() );
      
      assertEquals( 2.0f, e1.get( LEFT ).v, EPSILON );
      assertEquals( 3.7f, e1.get( RIGHT ).v, EPSILON );
      assertEquals( e0, e1 );
      
      e1.delete();
      e0.delete();
   }

   @Test
   public void testCloneShallow()
   {
      Entity e0 = new Entity( EXTENT );
      e0.get( LEFT ).v = 2.0f;
      e0.get( RIGHT ).v = 3.7f;
      e0.hide();
      
      Entity e1 = e0.clone( false );
      
      assertSame( e0.getTemplate(), e1.getTemplate() );
      assertSame( e0.get( LEFT ), e1.get( LEFT ) );
      assertSame( e0.get( RIGHT ), e1.get( RIGHT ) );
      assertFalse( e1.isVisible() );
      
      assertEquals( 2.0f, e1.get( LEFT ).v, EPSILON );
      assertEquals( 3.7f, e1.get( RIGHT ).v, EPSILON );
      assertEquals( e0, e1 );
      
      e1.delete();
      e0.delete();
   }
   
}
